File: cracking_the_coding_interview_6th/chapter17/cpp/17_03_v02.cpp
#include <chrono>
#include <random>
#include <string>
#include <vector>
using namespace std;
using ll = long long;
using hrc = chrono::high_resolution_clock;

vector<ll> pick_recursively(vector<ll> &org,
                            const ll m,
                            const ll i,
                            default_random_engine &generator) {
  if (i < m - 1)
    return {};
  else if (i == m - 1) {
    vector<ll> s(org.begin(), org.begin() + m);
    return s;
  } else {
    vector<ll> s = pick_recursively(org, m, i - 1, generator);
    uniform_int_distribution<ll> distribution(0, i);
    ll x = distribution(generator);
    if (x < m) s[x] = org[i];
    return s;
  }
}

vector<ll> gen_rand_set(vector<ll> &org, const ll m) {
  auto seed = hrc::now().time_since_epoch().count();
  default_random_engine generator{seed};
  vector<ll> res = pick_recursively(org, m, org.size() - 1, generator);
  return res;
}

void print_v(const vector<ll> &v) {
  for (ll i = 0; i < v.size(); ++i) {
    printf("%lld%c", v[i], i == v.size() - 1 ? '\n' : ' ');
  }
}

int main() {
  vector<ll> v{-9, -8, -7, -7, -6, -5, -5, -4, -3, -2, -2, -1, 0, 0, 0,
               1,  2,  3,  3,  4,  5,  5,  6,  6,  6,  7,  8,  8, 9, 9};
  for (ll i = 0; i < 5; ++i) {
    vector<ll> res = gen_rand_set(v, 5);
    print_v(res);
  }
}

File: cracking_the_coding_interview_6th/chapter17/python/17_03_v02.py
from typing import List
import random


def pick_recursively(org: List[int], m: int, i: int, rand: random) -> List[int]:
    if i < m - 1:
        return []
    elif i == m - 1:
        return org[:m].copy()
    else:
        s = pick_recursively(org, m, i - 1, rand)
        x = int(rand.uniform(0, i + 1))
        if x < m:
            s[x] = org[i]
        return s


def gen_rand_set(org: List[int], m: int) -> List[int]:
    rand = random
    return pick_recursively(org, m, len(org) - 1, rand)


v = [-9, -8, -7, -7, -6, -5, -5, -4, -3, -2, -2, -1, 0, 0, 0,
     1,  2,  3,  3,  4,  5,  5,  6,  6,  6,  7,  8,  8, 9, 9]
for i in range(5):
    print(gen_rand_set(v, 5))
